server/src/instant/grouped_queue.clj

```diff
@@ -5,7 +5,42 @@
    (java.util.concurrent.atomic AtomicInteger)
    (clojure.lang PersistentQueue)))
 
-(def empty-q PersistentQueue/EMPTY)
+;; ----------- 
+;; inflight-queue
+
+(def persisted-q-empty PersistentQueue/EMPTY)
+
+(def inflight-queue-empty
+  {:pending persisted-q-empty
+   :working []})
+
+(defn inflight-queue-put [inflight-queue item]
+  (update inflight-queue :pending conj item))
+
+(defn inflight-queue-empty? [{:keys [pending working] :as _inflight-queue}]
+  (and (empty? pending) (empty? working)))
+
+(defn inflight-queue-pop [{:keys [pending working]}]
+  {:pending (pop pending)
+   :working (if-let [item (first pending)]
+              (conj working item)
+              working)})
+
+(defn inflight-queue-workset [{:keys [working]}]
+  working)
+
+(defn inflight-queue-workset-clear [inflight-queue]
+  (assoc inflight-queue :working []))
+
+(defn inflight-queue-peek-pending [{:keys [pending] :as _inflight-queue}]
+  (first pending))
+
+(defn inflight-queue-drain [{:keys [pending working]}]
```

Comment:
Might be a good idea to have a `:max-items` field that will limit the number of items that can be moved from pending into working.

Reply:
Good idea! 

I went ahead and refactored like so: 

`(inflight-queue-reserve n iq)`

Using this to take `1` and to take `max-items` from pending

`(inflight-queue-reseve-all iq)` 

This takes all from pending. I could have had a signifier like `:all ` in inflight-queue-reserve, but thought separate functions would be logically clearer